src/main/kotlin/com/openlattice/indexing/IndexingService.kt

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (C) 2019. OpenLattice, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * You can contact the owner of the copyright at support@openlattice.com
+ *
+ *
+ */
+
+package com.openlattice.indexing
+
+import com.google.common.util.concurrent.ListeningExecutorService
+import com.hazelcast.core.HazelcastInstance
+import com.hazelcast.core.IMap
+import com.openlattice.admin.indexing.IndexingState
+import com.openlattice.conductor.rpc.ConductorElasticsearchApi
+import com.openlattice.data.storage.IndexingMetadataManager
+import com.openlattice.data.storage.PostgresEntityDataQueryService
+import com.openlattice.edm.EntitySet
+import com.openlattice.hazelcast.HazelcastMap
+import com.openlattice.hazelcast.HazelcastQueue
+import com.openlattice.hazelcast.processors.UUIDKeyToUUIDSetMerger
+import com.openlattice.rhizome.hazelcast.DelegatedUUIDSet
+import org.slf4j.LoggerFactory
+import java.util.*
+import java.util.concurrent.locks.ReentrantLock
+
+private val logger = LoggerFactory.getLogger(IndexingService::class.java)
+
+/**
+ *
+ * @author Matthew Tamayo-Rios &lt;matthew@openlattice.com&gt;
+ */
+class IndexingService(
+        private val backgroundIndexingService: BackgroundIndexingService,
+        private val executor: ListeningExecutorService,
+        hazelcastInstance: HazelcastInstance
+) {
+    private val entitySets: IMap<UUID, EntitySet> = hazelcastInstance.getMap(HazelcastMap.ENTITY_SETS.name)
+    private val indexingJobs = hazelcastInstance.getMap<UUID, DelegatedUUIDSet>(HazelcastMap.INDEXING_JOBS.name)
+    private val indexingQueue = hazelcastInstance.getQueue<UUID>(HazelcastQueue.INDEXING.name)
+    private val indexingLock = ReentrantLock()
+    private val indexingWorker = executor.submit {
+        while (true) {
+            try {
+                val entitySetId = indexingQueue.take()
+                indexingLock.lock()
+                val entitySet = entitySets.getValue(entitySetId)
```

Comment:
We should filter out linking entity sets or use BackgroundLinkingIndexingService in those cases

Reply:
On that note, how do linking entity sets differ from all other entity sets? I see that we some special cases around them throughout the codebase

Reply:
Linking entity sets contains no data, but they hold together/link together normal entity sets, which contain same kind of data. Right now, we only use it on general.person entity types, because the use case for it is to determine, if some of those individuals from different entity sets are actually the same person based on their property values, like name, SSN...etc. (see: RealtimeLinkingService, PersonProperties). We assign the same linking_ids(clusters) for these people.

Reply:
Okay cool, gotcha!

Reply:
I will add in an if statement to use background linking indexing service when it is a linking entity set.

Reply:
The linking case is still missing